client/app/bundles/course/assessment/containers/AssessmentForm/index.jsx

```diff
@@ -37,517 +38,734 @@ const styles = {
   }
 
+  // Shift bonus_end_at time
   if (
-    isTestCaseChosen(
-      values.use_public,
-      values.use_private,
-      values.use_evaluation,
-    )
+    newStartTime &&
+    oldStartTime &&
+    oldBonusTime &&
+    oldStartTime <= oldBonusTime
   ) {
-    errors.use_evaluation = translations.noTestCaseChosenError;
+    const nextBonusEndAt = new Date(
+      oldBonusTime + (newStartTime - oldStartTime),
+    );
+    setValue('bonus_end_at', nextBonusEndAt);
   }
-
-  return errors;
 };
 
-class AssessmentForm extends Component {
-  componentDidMount() {
-    const { dispatch, editing } = this.props;
-    // TODO: Shift the fetchTabs only when the selection menu is clicked on. This would
-    //  prevent unnecessary loading of the tabs every time the assessment form is loaded.
+const UpdatedAssessmentForm = (props) => {
+  const {
+    conditionAttributes,
+    disabled,
+    dispatch,
+    editing,
+    gamified,
+    folderAttributes,
+    initialValues,
+    modeSwitching,
+    onSubmit,
+    randomizationAllowed,
+    showPersonalizedTimelineFeatures,
+    tabs,
+  } = props;
+  const {
+    control,
+    handleSubmit,
+    setError,
+    setValue,
+    watch,
+    formState: { errors },
+  } = useForm({
+    defaultValues: initialValues,
+    resolver: yupResolver(validationSchema),
+  });
+  const autograded = watch('autograded');
+  const passwordProtected = watch('password_protected');
+
+  // Load all tabs if data is loaded, otherwise fall back to current assessment tab.
+  const loadedTabs = tabs || watch('tabs');
+
+  useEffect(() => {
     if (editing) {
       const failureMessage = (
         <FormattedMessage {...translations.fetchTabFailure} />
       );
       dispatch(fetchTabs(failureMessage));
     }
-  }
-
-  onStartAtChange = (_, newStartAt) => {
-    const {
-      start_at: startAt,
-      end_at: endAt,
-      bonus_end_at: bonusEndAt,
-      dispatch,
-    } = this.props;
-    const newStartTime = newStartAt && newStartAt.getTime();
-    const oldStartTime = startAt && new Date(startAt).getTime();
-    const oldEndTime = endAt && new Date(endAt).getTime();
-    const oldBonusTime = bonusEndAt && new Date(bonusEndAt).getTime();
-
-    // Shift end_at time
-    if (
-      newStartTime &&
-      oldStartTime &&
-      oldEndTime &&
-      oldStartTime <= oldEndTime
-    ) {
-      const newEndAt = new Date(oldEndTime + (newStartTime - oldStartTime));
-      dispatch(change(formNames.ASSESSMENT, 'end_at', newEndAt));
-    }
-
-    // Shift bonus_end_at time
-    if (
-      newStartTime &&
-      oldStartTime &&
-      oldBonusTime &&
-      oldStartTime <= oldBonusTime
-    ) {
-      const newBonusTime = new Date(
-        oldBonusTime + (newStartTime - oldStartTime),
-      );
-      dispatch(change(formNames.ASSESSMENT, 'bonus_end_at', newBonusTime));
-    }
-  };
-
-  renderEnableRandomizationField() {
-    const { submitting } = this.props;
+  }, []);
+
+  const renderPasswordFields = () => (
```

Comment:
@zhuhanming Is it better to declare this outside or inside the functional component?